### PR TITLE
Improved handling of authentication exceptions, which were undocumented

### PR DESCRIPTION
**Pull request by ee-usgs**
The underlying authentication throws a WebApplicationException for any type of authentication error,
even just bad credentials.  This is a type of RuntimeException, so does not need to be declared.
Since its not declared, users of the classes don't realize that they must handle this type
of exception as part of normal usage.

### CODE REVIEW

File: auth-web-utils/src/main/java/gov/usgs/cida/auth/ws/rs/service/AuthTokenService.java [hunk 1]
```diff
@@ -52,6 +53,8 @@ public Response getToken(
 			}
 		} catch(NotAuthorizedException ex) {
 			response = Response.status(Response.Status.UNAUTHORIZED).build();
+		} catch (WebApplicationException ex) {
+			response = Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
```
Comment (ee-usgs) [inline auth-web-utils/src/main/java/gov/usgs/cida/auth/ws/rs/service/AuthTokenService.java]: I don’t have a way to verity this the correct way to handle an internal server error here.  I assume its analogous to how the UNAUTHORIZED status is handled.

File: auth-client/src/main/java/gov/usgs/cida/auth/client/IAuthClient.java [hunk 2]
```diff
@@ -15,12 +16,29 @@
...
 	 * @param username
 	 * @param password
 	 * @return AuthToken
+	 * @throws WebApplicationException (A RuntimeException subclass) if the user
+	 *		provides invalid credentials or if there is an error during the login
+	 *		process.  Callers must handle this or risk this exception bubbling
+	 *		to the UI.
 	 */
-	public AuthToken getNewToken(String username, String password);
+	public AuthToken getNewToken(String username, String password) throws WebApplicationException;
```
Comment (ee-usgs) [inline auth-client/src/main/java/gov/usgs/cida/auth/client/IAuthClient.java]: Without the throws clause and docs, no one would know that a bad login will result in a RuntimeException.

Comment (thongsav-usgs) [inline auth-client/src/main/java/gov/usgs/cida/auth/client/IAuthClient.java]: This intentional. By convention, we are moving to unchecked exceptions. If you know a WebApplicationException == unauthorized, you should wrap it and throw the unchecked runtime NotAuthorizedException.

Documentation is preferred over a throws clause if you want to document this.

File: auth-client/src/main/java/gov/usgs/cida/auth/client/AuthClient.java [hunk 3]
```diff
@@ -61,7 +64,7 @@ public AuthClient(String authEndpoint) throws URISyntaxException {
 	 * {@inheritDoc}
 	 */
 	@Override
-	public AuthToken getNewToken(String username, String password) {
+	public AuthToken getNewToken(String username, String password) throws WebApplicationException {
```
Comment (thongsav-usgs) [inline auth-client/src/main/java/gov/usgs/cida/auth/client/AuthClient.java]: We are now preferring runtime exceptions. Mike gave a developer talk on this that sums up our strategy nicely but I don't have any resources that sum it up but can probably dig something up. To put it in really simple terms: Catch checked exceptions (especially those that aren't recoverable) and rethrow as a runtime exception, expect that the applications have a global error handling layer at the very top of the stack.

File: auth-client/src/main/java/gov/usgs/cida/auth/client/AuthClient.java [hunk 4]
```diff
@@ -83,6 +86,15 @@ public AuthToken getNewToken(String username, String password) {
 					ex.getResponse().getStatus() == Response.Status.UNAUTHORIZED.getStatusCode()) {
 				throw new NotAuthorizedException(ex.getMessage());
```
Comment (thongsav-usgs) [inline auth-client/src/main/java/gov/usgs/cida/auth/client/AuthClient.java]: Any exception that you know is likely caused by a not-authorized status should just throw this runtime exception.

**Comment by ee-usgs** [comment]
Hey @thongsav-usgs, I disagree w/ the approach here. Mike did give a presentation on Exceptions - I’m not sure his presentation automatically becomes accepted best practice, however.  I also don’t think Mike’s presentation applies here:  If an exception is an exceptional situation, it could be thrown as runtime (according to Mike).  However, in this case it is completely not exceptional and an expected as part of the normal process flow.

An API should be somewhat self-documenting:  If a method throws no exceptions, I should be able to assume I don't have to handle exceptions.  Worse, I can’t assume some other layer will catch my Runtime exception because, although Spring does catch it, it throws a bunch of stuff that ITSOT would not like to see on a page.

 A possible better approach would be to return a wrapper containing the token and a failure message or exception if the token couldn’t be created.  Barring that, at least declaring the exception would give the user a hint.

**Comment by thongsav-usgs** [comment]
I'm not citing mike's presentation as the reason we are doing it, I'm cited it a summary of the position taken, but it's probably not adequate since it's not actually documented.

I am fine with it being a checked exception, however "WebApplicationException" is too vague, and does  not allow any meaning compile-time action. There is no reason to pollute method signatures/apis with checked exceptions if they do not convey any context.

I have not dug deep enough to know where the WebApplicationException is spawned, but if you would like to make them checked we can come up with more well defined scenarios.

**Comment by ee-usgs** [comment]
I put in a separate PR using a checked requests here:  https://github.com/USGS-CIDA/cida-auth/pull/85

**Comment by ee-usgs** [comment]
Closing because it is superseded by #85.

**Pull request closed without merging**